File: packages/language-identification/android/src/main/java/io/capawesome/capacitorjs/plugins/mlkit/languageidentification/LanguageIdentificationPlugin.java
package io.capawesome.capacitorjs.plugins.mlkit.languageidentification;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import com.getcapacitor.Logger;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import io.capawesome.capacitorjs.plugins.mlkit.languageidentification.classes.options.IdentifyLanguageOptions;
import io.capawesome.capacitorjs.plugins.mlkit.languageidentification.classes.options.IdentifyPossibleLanguagesOptions;
import io.capawesome.capacitorjs.plugins.mlkit.languageidentification.classes.results.IdentifyLanguageResult;
import io.capawesome.capacitorjs.plugins.mlkit.languageidentification.classes.results.IdentifyPossibleLanguagesResult;
import io.capawesome.capacitorjs.plugins.mlkit.languageidentification.interfaces.NonEmptyResultCallback;
import io.capawesome.capacitorjs.plugins.mlkit.languageidentification.interfaces.Result;

@CapacitorPlugin(name = "LanguageIdentification")
public class LanguageIdentificationPlugin extends Plugin {

    public static final String TAG = "LanguageIdentification";

    public static final String ERROR_UNKNOWN_ERROR = "An unknown error occurred.";

    private LanguageIdentification implementation;

    @Override
    public void load() {
        implementation = new LanguageIdentification();
    }

    @PluginMethod
    public void identifyLanguage(PluginCall call) {
        try {
            IdentifyLanguageOptions options = new IdentifyLanguageOptions(call);
            NonEmptyResultCallback<IdentifyLanguageResult> callback = new NonEmptyResultCallback<>() {
                @Override
                public void success(@NonNull IdentifyLanguageResult result) {
                    resolveCall(call, result);
                }

                @Override
                public void error(@NonNull Exception exception) {
                    rejectCall(call, exception);
                }
            };

            implementation.identifyLanguage(options, callback);
        } catch (Exception exception) {
            rejectCall(call, exception);
        }
    }

    @PluginMethod
    public void identifyPossibleLanguages(PluginCall call) {
        try {
            IdentifyPossibleLanguagesOptions options = new IdentifyPossibleLanguagesOptions(call);
            NonEmptyResultCallback<IdentifyPossibleLanguagesResult> callback = new NonEmptyResultCallback<>() {
                @Override
                public void success(@NonNull IdentifyPossibleLanguagesResult result) {
                    resolveCall(call, result);
                }

                @Override
                public void error(@NonNull Exception exception) {
                    rejectCall(call, exception);
                }
            };

            implementation.identifyPossibleLanguages(options, callback);
        } catch (Exception exception) {
            rejectCall(call, exception);
        }
    }

    private void rejectCall(@NonNull PluginCall call, @NonNull Exception exception) {
        String message = exception.getMessage();
        if (message == null) {
            message = ERROR_UNKNOWN_ERROR;
        }
        Logger.error(TAG, message, exception);
        call.reject(message);
    }

    private void resolveCall(@NonNull PluginCall call, @Nullable Result result) {
        if (result == null) {
            call.resolve();
        } else {
            call.resolve(result.toJSObject());
        }
    }
}

File: packages/language-identification/ios/Plugin/LanguageIdentificationPlugin.swift
import Foundation
import Capacitor

/**
 * Please read the Capacitor iOS Plugin Development Guide
 * here: https://capacitorjs.com/docs/plugins/ios
 */
@objc(LanguageIdentificationPlugin)
public class LanguageIdentificationPlugin: CAPPlugin {
    public static let tag = "LanguageIdentification"

    private var implementation: LanguageIdentification?

    override public func load() {
        implementation = LanguageIdentification()
    }

    @objc func identifyLanguage(_ call: CAPPluginCall) {
        do {
            let options = try IdentifyLanguageOptions(call)

            implementation?.identifyLanguage(options, completion: { result, error in
                if let error = error {
                    self.rejectCall(call, error)
                } else {
                    self.resolveCall(call, result)
                }
            })
        } catch {
            self.rejectCall(call, error)
        }
    }

    @objc func identifyPossibleLanguages(_ call: CAPPluginCall) {
        do {
            let options = try IdentifyPossibleLanguagesOptions(call)

            implementation?.identifyPossibleLanguages(options, completion: { result, error in
                if let error = error {
                    self.rejectCall(call, error)
                } else {
                    self.resolveCall(call, result)
                }
            })
        } catch {
            self.rejectCall(call, error)
        }
    }

    private func rejectCall(_ call: CAPPluginCall, _ error: Error) {
        CAPLog.print("[", LanguageIdentificationPlugin.tag, "] ", error)
        call.reject(error.localizedDescription)
    }

    private func resolveCall(_ call: CAPPluginCall, _ result: Result?) {
        if let result = result?.toJSObject() as? JSObject {
            call.resolve(result)
        } else {
            call.resolve()
        }
    }
}
